locator/tablets.cc

```diff
@@ -387,6 +387,60 @@ const tablet_transition_info* tablet_map::get_tablet_transition_info(tablet_id i
     return &i->second;
 }
 
+const tablet_replica_set& tablet_map::get_replicas_for_reading(const tablet_id tablet) const {
+    auto&& info = get_tablet_transition_info(tablet);
+    auto&& replicas = std::invoke([&]() -> const tablet_replica_set& {
+        if (!info) {
+            return get_tablet_info(tablet).replicas;
+        }
+        switch (info->reads) {
+            case read_replica_set_selector::previous:
+                return get_tablet_info(tablet).replicas;
+            case read_replica_set_selector::next: {
+                return info->next;
+            }
+        }
+        on_internal_error(tablet_logger, format("Invalid replica selector {}", static_cast<int>(info->reads)));
+    });
+    return replicas;
+}
+
+const tablet_replica_set& tablet_map::get_replicas_for_writing(const tablet_id tablet) const {
+    auto* info = get_tablet_transition_info(tablet);
+    auto&& replicas = std::invoke([&]() -> const tablet_replica_set& {
+        if (!info) {
+            return get_tablet_info(tablet).replicas;
+        }
+        switch (info->writes) {
+            case write_replica_set_selector::previous:
+                [[fallthrough]];
+            case write_replica_set_selector::both:
+                return get_tablet_info(tablet).replicas;
+            case write_replica_set_selector::next: {
+                return info->next;
+            }
+        }
+        on_internal_error(tablet_logger, format("Invalid replica selector {}", static_cast<int>(info->writes)));
+    });
+    return replicas;
+}
+
+const std::optional<tablet_replica>* tablet_map::get_pending_replica(const tablet_id tablet) const {
```

Comment:
do we really need to return pointer to optional in this case? as this looks a bit strange ... I know that there could be a RVO issue with the optional, but I'd prefer readability in this case - the `optional<>` already accounts for the "empty result" case and it looks like we don't need to distinguish between pointer null or optional empty (and the only place where this is used needs to do both checks then)

Reply:
My intention was to return the exact object that's stored in tablets, that's why I used a pointer. But agree, it might be confusing.

Reply:
In the new version it returns `const tablet_replica*`, to avoid unnecessary memory copying.